docs/csharp/language-reference/keywords/lock-statement.md

```diff
@@ -11,76 +11,69 @@ ms.assetid: 656da1a4-707e-4ef6-9c6e-6d13b646af42
+The `lock` statement obtains the mutual-exclusion lock for a given object, executes a statement block, and then releases the lock.
+
+The `lock` statement of the form
 
 ```csharp
-class Account
+lock (x)
 {
-    decimal balance;
-    private Object thisLock = new Object();
-
-    public void Withdraw(decimal amount)
-    {
-        lock (thisLock)
-        {
-            if (amount > balance)
-            {
-                throw new Exception("Insufficient funds");
-            }
-            balance -= amount;
-        }
-    }
+    // Your code...
 }
 ```
 
-For more information, see [Thread Synchronization](../../programming-guide/concepts/threading/thread-synchronization.md).
-
-## Remarks
-
-The `lock` keyword ensures that one thread does not enter a critical section of code while another thread is in the critical section. If another thread tries to enter a locked code, it will wait, block, until the object is released.
+where `x` is an expression of a [reference type](reference-types.md) is precisely equivalent to
 
-The section [Threading](../../programming-guide/concepts/threading/index.md) discusses threading.
+```csharp
+bool __lockWasTaken = false;
+try
+{
+    System.Threading.Monitor.Enter(x, ref __lockWasTaken);
+    // Your code...
+}
+finally
+{
+    if (__lockWasTaken) System.Threading.Monitor.Exit(x);
+}
+```
 
-The `lock` keyword calls <xref:System.Threading.Monitor.Enter%2A> at the start of the block and <xref:System.Threading.Monitor.Exit%2A> at the end of the block. A <xref:System.Threading.ThreadInterruptedException> is thrown if <xref:System.Threading.Thread.Interrupt%2A> interrupts a thread that is waiting to enter a `lock` statement.
+except that `x` is only evaluated once.
 
-In general, avoid locking on a `public` type, or instances beyond your code's control. The common constructs `lock (this)`, `lock (typeof (MyType))`, and `lock ("myLock")` violate this guideline:
+Since the code uses a [try...finally](try-finally.md) block, the lock is released even if an exception is thrown within the body of a `lock` statement.
 
-- `lock (this)` is a problem if the instance can be accessed publicly.
+While a lock is held, the same thread can again obtain and release the lock. However, any other thread is blocked from obtaining the lock and waits until the lock is released.
 
-- `lock (typeof (MyType))` is a problem if `MyType` is publicly accessible.
+You can't use the [await](await.md) keyword in the body of a `lock` statement.
 
-- `lock("myLock")` is a problem because any other code in the process using the same string, will share the same lock.
+## Remarks
 
-Best practice is to define a `private` object to lock on, or a `private static` object variable to protect data common to all instances.
+Avoid locking on publicly accessible instances, as it might result in lock contention or deadlock. Consider the following constructs:
```

Comment:
Using "publicly accessible" in relation to object instances is potentially confusing, especially that one of the cases below also uses that in relation to types. Though I'm not really sure what terminology works best here. Maybe it should say something like:

Avoid using the same lock object instance for different shared resources, as it might result in deadlock or lock contention. Use a dedicated object instance (`private readonly object listLock = new object()`) or another instance that is unlikely to be used as a lock object by unrelated parts of the code. This in general precludes the use of `this` (potentially used as a lock by the callers), `Type` instances (be it obtained through `typeof` or reflection) or string instances, including string literals.

Might be a bit convoluted but anyway, something along those lines.

Reply:
@mikedn thanks, this comment helped a lot.